src/SFML/Graphics/Shader.cpp

```diff
@@ -156,11 +210,50 @@ Shader::CurrentTextureType Shader::CurrentTexture;
 
 
 ////////////////////////////////////////////////////////////
+struct Shader::UniformBinder : private sf::NonCopyable
+{
+    ////////////////////////////////////////////////////////////
+    /// \brief Constructor: set up state before uniform is set
+    ///
+    ////////////////////////////////////////////////////////////
+    UniformBinder(Shader* shader, const std::string& name) :
+    program(0),
+    location(-1)
+    {
+        if (shader->m_shaderProgram)
```

Comment:
What is your definition for a resource? For me, an object that actually stores and even changes the bound shader program is just as much a resource as Texture or Shader.

The constructor and destructor in GlResource are there for a reason: to make sure that the shared context is available before anybody calls ensureContext. We already had 2 cases of bugs which involved ensureContext being called before any GlResource was constructed. Working around having to construct a GlResource to access its static member functions is just another potential source of bugs.

The constructor and destructor don't even do that much. If the shared context already exists, they simply lock and unlock a mutex.